### DISCUSSION TITLE
Eleventy and Baserow

### DISCUSSION THREAD

**Discussion by xplosionmind** [11ty/eleventy#2430] [Q&A]
Hi to you all! I am just getting started with learning JasvaScript, but I am a heavy Eleventy user. Still, I have much more to learn.

In particular, I was wondering if somebody could help me in figuring out how to automatically fetch data from [Baserow](https://baserow.io) when building my Eleventy website. Is there a way I can use data stored remotely instead of relying on files stored in the `data` folder?

I found very few tutorials about using Baserow API, and knowing very, very little JavaScript I cannot really get my head around it.

Any link to insightful resources or any help would be very much appreciated!

**Comment by pdehaan** [comment]
I've never used Baserow, but&hellip;
1. I'd check npm to see if they have some official (or unofficial) module you can use. I did a quick search and didn't see one, but I did find some Gatsby module, https://www.npmjs.com/package/gatsby-source-baserow. You could see how they're talking w/ Baserow and try something similar w/ Eleventy.
2. Check their docs and see if they have an API and then fetch the content using something like https://www.npmjs.com/package/axios
3. Use something like https://www.11ty.dev/docs/plugins/fetch/ which does some really useful fetching+caching, but sometimes needs a bit more configuration than something simpler like axios/got/fetch. But the benefit is that caching is a lot nicer if you have strict API limits (like GitHub's 60 requests per hour, or whatever the exact numbers are).

Then I typically would use a global data file to fetch the data from the third party API. Usually works well, unless you need to pass specific data to an API (which global data files don't generally support). So depending on how you're using Baserow, you might need to wrap in a custom shortcode or filter instead. 🤷 
Or another thing I've done is have another Node script which fetches+transforms the data and then manually writes it into the _data/ directory. For example, some data might only change daily/weekly (or just take minutes to fetch and process), so I can just manually fetch it when I need to instead of on every single build/reload.

**Reply by Ryuno-Ki** [reply]
Especially these two lines appear to fetch the content:

https://github.com/mabhub/gatsby-source-baserow/blob/cd2af2710d2f3d8b07fdaa5a9ceeee8dade119b8/gatsby-node.js#L65-L66

Their API appear to be documented at https://baserow.io/docs/apis/rest-api

Having a separate script is preferable in my experience.

**Reply by pdehaan** [reply]
Yeah, and those two functions are defined in https://github.com/mabhub/gatsby-source-baserow/blob/cd2af2710d2f3d8b07fdaa5a9ceeee8dade119b8/lib.js.
That's actually a pretty interesting use of a generator to fetch paged results. I might have to ~steal~ borrow that approach in the future.

**Reply by xplosionmind** [reply]
Thanks a lot, @pdehaan and @Ryuno-Ki! Your help is very valuable but, unfortunately, I am not as good with JavaScript to perform anything you suggested me to do. Probably, considering my embarrassing little knowledge, the following option is more suitable for me:

> have another Node script which fetches+transforms the data and then manually writes it into the _data/ directory.

Could you help me with that? Do you have any examples? I am not good enough to write a script from scratch.

---

In any case, I would like to start trying and use the Baserow APIs. Ideally, I would create a global JavaScript file that fetches all of the data from Baserow. Do you have any more specific examples, other than the ones you already linked?

Thanks a lot!
❤️

**Reply by pdehaan** [reply]
I've never used Baserow or Airtable, so I don't have any examples. You might have better luck in their repos/forums.

The good news is, if you're going with an external script solution, you aren't limited to using Node; you could use Bash or Python or PHP or cURL or whatever else you want to consume the Baserow API. (For example, here's a Python library: https://niklasrosenstein.github.io/python-baserow-client/getting-started/, and they had a PHP library listed on https://baserow.io/docs/other/external-resources). I tried searching their repo for "nodejs" or examples at https://gitlab.com/bramw/baserow/-/issues/, but apparently I need to create a GitLab account and sign in just to search issues.

> _"You must sign in to search for specific terms."_

The less good news is, it's oddly difficult to find information on using the Baserow API on their website, or on The Internet(tm). Most examples/videos I found this morning just walked through the UI for entering data. I thought Baserow was an Airtable clone, but when I tried looking at the Airtable API, they want me to sign up just to view their docs apparently: https://airtable.com/api. 🤷

**Comment by TigersWay** [comment]
This seems to be the perfect example for [Quick Tip #009—Cache Data Requests](https://www.11ty.dev/docs/quicktips/cache-api-requests/)
- https://api.baserow.io/api/database/rows/table/<tableID>/?user_field_names=true and a simple Authorization token

Not very fast but working well.
